### DISCUSSION TITLE
3Gpp parameter modeling for Network Functions

### DISCUSSION THREAD

**Discussion by gvbalaji** [nephio-project/nephio#420] [Development]
Here we will discuss about how to model common, specific and customer extensions for 3gpp parameters for network functions.

**Comment by s3wong** [comment]
@arora-sagar and I had a discussion on Slack w.r.t. this topic. So, to recapture free5gc's needs:

`Nephio free5gc operator supports configuring of three NFs: AMF, SMF, and UPF. 

**UPF** only needs DNN (which is already in NFDeployment now); 
**AMF** and **SMF** both need PLMN(s); **SMF** additionally needs S-NSSAI parameters (sst, sd) for each slice
There are also some parameters (in **SMF**) that we aren’t setting for now: t3591 and t3592 timeouts
And **AMF** additionally has parameters for T3502, T3512, T3513, T3522, T3550, T3560, T3565, T3570, and some SCTP related config`

**Comment by arora-sagar** [comment]
I managed to define common-3gpp params for RAN and Core (I checked it should be enough for OAI, Free5GC, and Open5GS). 

```conf
plmnInfoList:
  - mcc: 001
    mcc: 01
    tac: 1
    nssai:
      - sst: 1
        sd: ffffff
        dnnInfoList:
          - dnn: internet
            session_type: IPV4
            dns: 8.8.8.8
      - sst: 2
        sd: ffffff
        dnnInfoList:
          - dnn: internet
            session_type: IPV4
            dns: 8.8.8.8
  - mcc: 208
    mnc: 99
    nssai:
      - sst: 1
        sd: ffffff
        dnnInfoList:
          - dnn: internet
            session_type: IPV4
            dns: 8.8.8.8
communicationModel: 
  - name: with_nrf
    nrfUri: http://nrf-nnrf:8000
```
The field `communicationModel` is inspired from https://forge.3gpp.org/rep/sa5/MnS/-/blob/Rel-16/yang-models/_3gpp-5g-common-yang-types.yang#L54 

I can define the CRD to verify that the inputs are correct.  

`PS`: The resource for Open5Gs I used is here --> https://github.com/openshift-telco/distributed-open5gs/tree/upf-only/5gcore/templates

I will make something similar for RAN like common-ran.

**Reply by henderiw** [reply]
With this proposal we need to rework the dnn function and as I said before we should adopt ipv4 and ipv6 according k8s terminology. This is what we did so far. See interface CRD(s) we have so far.

**Reply by arora-sagar** [reply]
@henderiw I did the changes, so we can get the IP pool from the DNN/interface.

**Comment by arora-sagar** [comment]
Should I also model the 3GPP parameters that are needed only by a specific NF? For example AMF/SMF/DU/CU-CP/CU-UP? This means I should create a config CRD for each of these or for R2 common configuration should be enough?

**Reply by s3wong** [reply]
Are they the same across all the vendors? For example, as we discussed, PLMNs are needed for both AMF and SMF for free5gc, but perhaps only for AMF for OAI?

TBH, having extra information pushed to the workload cluster isn't deadly (not enough is); but if the parameters are organized in a way where it could be delivered in multiple CRs vs just one giant one, that could make it more flexible.

**Reply by henderiw** [reply]
I believe we should model them as well.

**Reply by arora-sagar** [reply]
OKay I did you can check @henderiw and @s3wong

**Comment by arora-sagar** [comment]
Here is the model for AMF

```conf
managementInfo:
  vendorName: example
  swVersion: develop
amfName: amf
sbi:
  fqdn: amf.example.org
  scheme: http/https
serviceNameList:
  - namf-comm
  - namf-evts
  - namf-mt
  - namf-loc
  - namf-oam
servedGuamiList:
  - plmnId:
      mcc: 001
      mnc: 01
    amfId:
      amfRegionId: 1
      amfSetId: 1
      amfPointer: 1
  - plmnId:
      mcc: 208
      mnc: 99
    amfId:
      amfRegionId: 1
      amfSetId: 1
      amfPointer: 1
security:
    integrityOrder: 
      - NIA2 
      - NIA1 
      - NIA0 
    cipheringOrder: 
      - NEA0
      - NEA1
      - NEA2
```

**Comment by arora-sagar** [comment]
Here is the smf model

```conf
managementInfo:
  vendorName: example
  swVersion: develop
smfName: smf
sbi:
  fqdn: smf.example.org
  scheme: http/https
serviceNameList:
  - nsmf-pdusession
  - nsmf-event-exposure
  - nsmf-oam
```

UPF doesn't need anything specific lets say for the moment. So we can only model these 3GPP specific CRD. Vendor-specific information can be in a CR and the vendor will take care of verifying the information. 

Btw the `swVersion` field can be used to define imagetag.

**Comment by arora-sagar** [comment]
I haven't defined, `apiVersion` and `kind` and yet. So we have to define that. I just defined the parameters if that is fine. Then the kinds and apiVersion are easy to choose.

**Comment by henderiw** [comment]
in general would it not be better to make a PR with the proposal and we can comment on the code?
This is where we need to end up anyhow, so maybe better to add a PR and put the comments there.

My 2 cents.

**Reply by arora-sagar** [reply]
Okay I will do that

**Comment by gvbalaji** [comment]
3gpp yang models
https://forge.3gpp.org/rep/sa5/MnS/-/tree/Rel-18/yang-models
https://forge.3gpp.org/rep/all/5G_APIs
